Singular cloud mode documentation updates (#670)
rudderlabs/rudderstack-docs: docs/destinations/analytics/singular/singular-cloud-mode.mdx

<div class="warningBlock">
RudderStack considers an event as a session event only if it is specified in the dashboard settings or if it one of the following three <Link to="/rudderstack-api/api-specification/application-lifecycle-events-spec/">lifecycle events</Link>:
<ul>
  <li>Application Installed</li>
  <li>Application Opened</li>
  <li>Application Updated</li>
</ul>
</div>

<div class="infoBlock">
RudderStack automatically tracks the above three lifecycle events if lifecycle event tracking is enabled.
</div>

### Supported mappings

This section lists the mappings of the RudderStack event properties to the relevant Singular fields.

The following table lists the mapping of the attributes **automatically captured** by RudderStack for the **mobile platforms** (Android and iOS):

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----| 
| `context.os.name` | `p` | Required | The source platform (Android or iOS). |
| `context.app.namespace` | `i` | Required | The package name (Android) or bundle ID (iOS) of your app. |
| `context.app.version` | `app_v` | Required | The app version. |

> | `requestIP` | `ip` | Required | The user's IP address. |

Reviewer: ```suggestion
| `context.ip / request_ip` | `ip` | Required | The user's IP address. |
```

| `context.os.version` | `ve` | Required | The device OS version at session time. |
| `context.device.model` | `mo` | Required | The device model. This parameter must be used with the `ma` parameter. |

> | `context.device.manufacturer` | `mo` | Required | The make of the device hardware. This parameter must be used with the `mo` parameter. |

Reviewer: ```suggestion
| `context.device.manufacturer` | `ma` | Required | The make of the device hardware. This parameter must be used with the `mo` parameter. |
```

> | `context.locale` | `lc` | Required | The device's IETF local tag using two-lettered language and country code, separated by an underscore. |

Author: ```suggestion
| `context.locale1` | `lc` | Required | The device's IETF local tag using two-lettered language and country code, separated by an underscore. |
```

> | `context.device.advertisingId` | `idfa` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. This is applicable for iOS apps only. |

Reviewer: Can we have both advertisingId and device.id for android and iOS together in one block but with their exclusive descriptions to avoid repitition in mapping?

> | `context.device.advertisingId` | `idfa` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. This is applicable for iOS apps only. |

Reviewer: ```suggestion
| `context.device.advertisingId` | `idfa` | Required | The raw [`[advertising ID](https://developer.apple.com/library/ios/documentation/AdSupport/Reference/ASIdentifierManager_Ref/index.html)`in upper case with dashes. This is applicable for iOS apps only. |
```

| `context.device.id` | `idfv` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. This is applicable for iOS apps only. |
| `context.device.advertisingId`| `aifa` | Required | Applicable for Android apps only. This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. |
| `context.device.id` | `andi` | Required | The raw [`Android ID`](http://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID) in lower case. This is applicable for Android apps only. This is required only when Android Advertising ID is unavailable on the device. |
| `context.app.build` | `bd` | Required | The device build (URL encoded). |

> | `context.device.adTrackingEnabled` | `dnt` | Required | Pass `1` if do not track is enabled, else pass `0`. This is automatically captured if you pass the advertising ID to the SDK. |

Reviewer: ```suggestion
| `context.device.adTrackingEnabled` | `dnt` | Required | Pass `true` if do not track is disabled, else pass `false`. This is automatically captured if you pass the advertising ID to the SDK. |
```

| `context.device.token` | `gcm` | Optional | The Google Cloud Messaging Device Token. It is required for uninstall tracking in Android (legacy devices). |
| `context.device.token` | `apns_token` | Optional | The Apple Push Notification Service Device Token. It is required for uninstall tracking in iOS. |
| `context.device.advertisingId`| `aifa` | Required | Applicable for Android apps only. This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. |

## Custom events

RudderStack sends all the other events other than the session events as custom events via Singular's [`evt`](https://s2s.singular.net/api/v1/evt) endpoint.

The following table lists the mapping of the attributes **automatically captured** by RudderStack for the **mobile platforms** (Android and iOS):

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----| 
| `context.os.name` | `p` | Required | The source platform (Android or iOS). |
| `context.app.namespace` | `i` | Required | The package name (Android) or bundle ID (iOS) of your app. |

> | `requestIP` | `ip` | Required | The user's IP address. |

Author: ```suggestion
| `request_ip` / `context.ip` | `ip` | Required | The user's IP address. |
```

The following table lists the mapping of the attributes **must be passed via the event properties**:

<div class="warningBlock">
These properties are not persisted in the SDK and must be passed with every event.
</div>

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----|
| `properties.install_ref` | `install_ref` | Required | The Google Install Referrer Information. |
| `properties.referring_application` | `install_source` | Required | The install source package name in Android. Use [`getInitiatingPackageName()`](https://developer.android.com/reference/android/content/pm/InstallSourceInfo#getInitiatingPackageName()) to retrieve this. |
| `properties.install_receipt` | `install_receipt` | Required | The receipt received from the install. To retreive this, follow the [iOS Install Receipt](https://support.singular.net/hc/en-us/articles/360037640812#Reference_Retrieving_the_iOS_Install_Receipt) guide. |
| `properties.asid` | `asid` | Required | The App Set ID for Android v12+ devices. |
| `properties.url` | `openui` | Required | If the app is opened via a deep link/universal link, the value of the encoded deep link URL. |

> | `context.device.attStatus` | `att_authorization_status` | Required | The [App Tracking Transparency authorization status](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanager/authorizationstatus). |

Reviewer: ```suggestion
| `context.device.attTrackingStatus` | `att_authorization_status` | Required | The [App Tracking Transparency authorization status](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanager/authorizationstatus). |
```

| `context.device.advertisingId`| `aifa` | Required | Applicable for Android apps only. This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. |

<div class="infoBlock">
RudderStack supports <code class="inline-code">fcm</code> over  <code class="inline-code">gcm</code> (legacy) for mapping two values to one key.
</div>

## Custom events

RudderStack sends all the other events other than the session events as custom events via Singular's [`evt`](https://s2s.singular.net/api/v1/evt) endpoint.

The following table lists the mapping of the attributes **automatically captured** by RudderStack for the **mobile platforms** (<Link to="/stream-sources/rudderstack-sdk-integration-guides/rudderstack-android-sdk/">Android</Link> and <Link to="/stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/">iOS</Link>):

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----| 
| `context.os.name` | `p` | Required | The source platform (Android or iOS). |
| `context.app.namespace` | `i` | Required | The package name (Android) or bundle ID (iOS) of your app. |

> | `request_ip` | `ip` | Required | The user's IP address. |

Reviewer: ```suggestion
| `context.ip`/`request_ip` | `ip` | Required | The user's IP address. |
```

> | `context.device.advertisingId` | `idfa` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. This is applicable for iOS apps only. |

Reviewer: Same as in session events, can we have Ids for android and iOS together but with descriptions?

Reviewer: to avoid repetition in the table.

| `context.device.advertisingId`| `aifa` | Required | This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. This is applicable for Android apps only.  |
| `context.device.id` | `andi` | Required | The raw [`Android ID`](http://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID) in lower case. This is applicable for Android apps only and is required only when the Android Advertising ID is unavailable on the device. |
| `context.os.version` | `ve` | Required | The device OS version at session time. |
| <Link to="/rudderstack-api/api-specification/rudderstack-spec/common-fields/#clock-skew-considerations"><code class="inline-code">timestamp</code></Link> | `utime` | Optional | The session time (in UNIX time). |

<div class="infoBlock">
Singular prefers <code class="inline-code">aifa</code> over <code class="inline-code">andi</code> (in Android) and <code class="inline-code">idfa</code> over <code class="inline-code">idfv</code> (in iOS)
</div>

> The following table lists the mapping of the attributes **must be passed via the event properties**:

Reviewer: Can we have the RudderStack properties mentioned here as properties.{propertyname} for each entry in this table?

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----|
| `event` | `n` | Required | The name of the event. **This is user-defined**. |

> | `context.device.attStatus` | `att_authorization_status` | Required | The [App Tracking Transparency authorization status](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanager/authorizationstatus). |

Reviewer: ```suggestion
| `context.device.attTrackingStatus` | `att_authorization_status` | Required | The [App Tracking Transparency authorization status](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanager/authorizationstatus). |
```

| `userId` | `custom_user_id` | Optional | The user ID passed through the `identify` call. |
| `properties.skan_conversion_value` | `skan_conversion_value` | Optional | The latest `SkAdNetwork` value at the time of the session notification. |
| `properties.skan_first_call_timestamp` | `skan_first_call_timestamp` | Optional | UNIX timestamp of the first call made to the `SkAdNetwork` API. |
| `properties.skan_last_call_timestamp` | `skan_last_call_timestamp` | Optional | UNIX timestamp of the last call made to the `SkAdNetwork` API at the time of the session notification. |
| `properties.eventAttributes` | `e` | Optional | The custom event attributes in JSON format. You need to pass these with every event as they are not persisted in the SDK. |

> | `properties.is_revenue_event` | `is_revenue_event` | Optional | Determines if an event is a revenue event. You need to pass this through the properties with every event as it is not persisted in the SDK. |

Reviewer: We need to specify a callout related to this, and mention this:

If any one out of `value`, `revenue`, `total` is set, we'll consider the event as revenue event, unless explicitly mentioned by `is_revenue_event` property

| `context.os.version` | `ve` | Required | The device OS version at session time. |
| `context.device.model` | `mo` | Required | The device model. This parameter must be used with the `ma` parameter. |
| `context.device.manufacturer` | `ma` | Required | The make of the device hardware. This parameter must be used with the `mo` parameter. |
| `context.locale` | `lc` | Required | The device's IETF local tag using two-lettered language and country code, separated by an underscore. |
| `context.device.id` | `idfv` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. **This is applicable for iOS apps only**. |
| `context.device.id` | `andi` | Required | The raw [`Android ID`](http://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID) in lower case. **This is applicable for Android apps only** and is required only when the Android advertising ID is unavailable on the device. |
| `context.app.build` | `bd` | Required | The device build (URL encoded). |

> | `context.device.adTrackingEnabled` | `dnt` | Required | Pass `true` if do not track is enabled, else pass `false`. This is automatically captured if you pass the advertising ID to the SDK. |

Reviewer 2: `context.device.adTrackingEnabled` when set to `true` this means add tracking will be enabled so `dnt` (do not track) should be disabled i.e. its value will be set to 0 (i.e. disabled) and when `adTrackingEnabled` is `false` , `dnt` will be set to `1` (i.e. enabled)

Author: I've fixed the wording for this - @utsabc

| `context.device.advertisingId` | `idfa` | Required | The raw [advertising ID](https://developer.apple.com/library/ios/documentation/AdSupport/Reference/ASIdentifierManager_Ref/index.html) in upper case with dashes. **This is applicable for iOS apps only**. |
| `context.device.advertisingId`| `aifa` | Required | This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. **This is applicable for Android apps only**. |

<div class="infoBlock">
For more information on setting the device token, refer to the relevant SDK documentation:
<ul>
  <li><Link to="/stream-sources/rudderstack-sdk-integration-guides/rudderstack-android-sdk/#setting-the-android-device-token">Android</Link></li>
  <li><Link to="stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/#setting-the-device-token">iOS</Link></li>
</ul>
</div>

<div class="infoBlock">
RudderStack supports only <code class="inline-code">fcm</code> for mapping two values to one key.
</div>

## Custom events

> RudderStack sends all the other events other than the session events as custom events via Singular's [`evt`](https://s2s.singular.net/api/v1/evt) endpoint.

Reviewer 3: ```suggestion
RudderStack sends all the events other than the session events as custom events via Singular's [`evt`](https://s2s.singular.net/api/v1/evt) endpoint.
```

> 

Author: agreed. made this change.

| `context.device.advertisingId`| `aifa` | Required | This is the lower case raw [advertising ID](http://support.google.com/googleplay/android-developer/answer/6048248?hl=en) with dashes. **This is applicable for Android apps only**.  |
| `context.device.id` | `idfv` | Required | The raw [`IdentifierForVendor`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIDevice_Class/Reference/UIDevice.html#//apple_ref/occ/instp/UIDevice/identifierForVendor) in upper case with dashes. **This is applicable for iOS apps only**. |
| `context.device.id` | `andi` | Required | The raw [`Android ID`](http://developer.android.com/reference/android/provider/Settings.Secure.html#ANDROID_ID) in lower case. **This is applicable for Android apps only** and is required only when the Android Advertising ID is unavailable on the device. |
| `context.os.version` | `ve` | Required | The device OS version at session time. |
| <Link to="/rudderstack-api/api-specification/rudderstack-spec/common-fields/#clock-skew-considerations"><code class="inline-code">timestamp</code></Link> | `utime` | Optional | The session time (in UNIX time). |

<div class="infoBlock">

> Singular prefers <code class="inline-code">aifa</code> over <code class="inline-code">andi</code> (in Android) and <code class="inline-code">idfa</code> over <code class="inline-code">idfv</code> (in iOS)

Reviewer 3: ```suggestion
Singular prefers <code class="inline-code">aifa</code> over <code class="inline-code">andi</code> (in Android) and <code class="inline-code">idfa</code> over <code class="inline-code">idfv</code> (in iOS).
```

  <li><Link to="stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/#setting-the-device-token">iOS</Link></li>
</ul>
</div>

<div class="infoBlock">
RudderStack supports only <code class="inline-code">fcm</code> for mapping two values to one key.
</div>

## Custom events

RudderStack sends all the events other than the session events as custom events via Singular's [`evt`](https://s2s.singular.net/api/v1/evt) endpoint.

### Supported mappings

The following table lists the mapping of the attributes **automatically captured** by RudderStack for the **mobile platforms** (<Link to="/stream-sources/rudderstack-sdk-integration-guides/rudderstack-android-sdk/">Android</Link> and <Link to="/stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/">iOS</Link>):

| RudderStack property | Singular attribute | Presence | Description |
| :----| :----| :----| :----| 
| `context.os.name` | `p` | Required | The source platform (Android or iOS). |
| `context.app.namespace` | `i` | Required | The package name (Android) or bundle ID (iOS) of your app. |

> | `context.ip` / `request_ip` (in that order) | `ip` | Required | The user's IP address. |

Reviewer 2: ```suggestion
| `context.ip` / `request_ip` (in same order) | `ip` | Required | The user's IP address. |
```

| `properties.purchase_receipt` | `purchase_receipt` | Optional | The receipt received from a purchase. |
| `properties.product_id`/`properties.sku` | `purchase_product_id` | Optional | The product SKU identifier. |
| `properties.orderId` / `properties.purchase_transaction_id` (in that order) | `purchase_transaction_id` | Optional | The transaction identifier. |

<div class="infoBlock">
If you set any one out of the <code class="inline-code">value</code>, <code class="inline-code">revenue</code>, or <code class="inline-code">total</code> properties, RudderStack automatically considers the event as a revenue event, unless it is explicitly mentioned by the <code class="inline-code">is_revenue_event</code> property.
</div>

A few important considerations in case of custom events are listed below:

- RudderStack takes the user agent from `context.userAgent` for Android and from the event properties in case of iOS.
- RudderStack supports `fcm` over `gcm` (legacy) for mapping two values to one key.
- RudderStack stores the extra attributes passed in the custom event in Singular's `e` field.

> ## Contact us

Reviewer 2: @yashasvibajpai @ameypv-rudder Can we add how to verify event delivery Via live events and reports in faq section

Reviewer: For FAQ, @ameypv-rudder We can mention something on the lines of:
For realtime event verification, use SDK Console, while for more granular reports, we can make use of Attribution > Export logs (It won't be realtime though)

A few important considerations in case of custom events are listed below:

- RudderStack takes the user agent from `context.userAgent` for Android and from the event properties in case of iOS.

> - RudderStack supports only `fcm` for mapping two values to one key.

Reviewer: RudderStack only supports fcm. 
for mapping two values to one key. -> this is redundant since we have 1 value and 1 key

Author: Updated this information.
